File: tests/coretests/Stocks/Services/NumberAnalysisTests.cs
using core.Stocks.Services.Analysis;
using Xunit;

namespace coretests.Stocks.Services
{
    public class NumberAnalysis_PercentChangesTests
    {
        private DistributionStatistics _result;

        public NumberAnalysis_PercentChangesTests()
        {
            _result = NumberAnalysis.PercentChanges(new decimal[] { 1, 2, 3, 4, 5 });
        }

        [Fact]
        public void MeanIsCorrect() => Assert.Equal(0.52m, _result.mean);

        [Fact]
        public void StdDevIsCorrect() => Assert.Equal(0.34m, _result.stdDev);

        [Fact]
        public void MinIsCorrect() => Assert.Equal(0.25m, _result.min);

        [Fact]
        public void MaxIsCorrect() => Assert.Equal(1m, _result.max);

        [Fact]
        public void MedianIsCorrect() => Assert.Equal(0.50m, _result.median);

        [Fact]
        public void SkewnessIsCorrect() => Assert.Equal(0.54m, _result.skewness);

        [Fact]
        public void KurtosisIsCorrect() => Assert.Equal(-1.88m, _result.kurtosis);

        [Fact]
        public void BucketsAreCorrect()
        {
            Assert.Equal(21, _result.buckets.Length);
            Assert.Equal(-10, _result.buckets[0].percentChange);
            Assert.Equal(0, _result.buckets[0].frequency);
            Assert.Equal(-9, _result.buckets[1].percentChange);
            Assert.Equal(0, _result.buckets[1].frequency);
            Assert.Equal(10, _result.buckets[^1].percentChange);
            Assert.Equal(4, _result.buckets[^1].frequency);
        }

        [Fact]
        public void CountIsCorrect() => Assert.Equal(4, _result.count);
    }
}

File: tests/coretests.fs/Stocks/Services/NumberAnalysisTests.fs
module coretests.fs.Stocks.Services.NumberAnalysisTests

open Xunit
open core.fs.Services.Analysis
open testutils


module PercentChangesTests =
    let percentChangeStatistics = PercentChangeAnalysis.calculate false [1m; 2m; 3m; 4m; 5m]

    [<Fact>]
    let ``Mean is correct``() = Assert.Equal(0.52m, percentChangeStatistics.mean, 2)

    [<Fact>]
    let ``StdDev is correct``() = Assert.Equal(0.34m, percentChangeStatistics.stdDev, 2)

    [<Fact>]
    let ``Min is correct``() = Assert.Equal(0.25m, percentChangeStatistics.min)

    [<Fact>]
    let ``Max is correct``() = Assert.Equal(1m, percentChangeStatistics.max)

    [<Fact>]
    let ``Median is correct``() = Assert.Equal(0.5m, percentChangeStatistics.median, 2)

    [<Fact>]
    let ``Skewness is correct``() = Assert.Equal(1.47m, percentChangeStatistics.skewness, 2)

    [<Fact>]
    let ``Kurtosis is correct``() = Assert.Equal(2.03m, percentChangeStatistics.kurtosis, 2)

    [<Fact>]
    let ``Buckets are correct``() =
        Assert.Equal(21, percentChangeStatistics.buckets.Length)
        
        // make sure there are four buckets with values assigned
        Assert.Equal(4, percentChangeStatistics.buckets |> Array.filter (fun x -> x.frequency <> 0) |> Array.length)
        
        // last bucket should be max
        let lastBucketValue = percentChangeStatistics.buckets |> Array.last |> _.value
        let lastBucketFrequency = percentChangeStatistics.buckets |> Array.last |> _.frequency
        
        Assert.True(percentChangeStatistics.max >= lastBucketValue)
        Assert.Equal(1, lastBucketFrequency)

    [<Fact>]
    let ``Count is correct``() = Assert.Equal(4m, percentChangeStatistics.count)
    
module RealBarsTests =
    
    let percentChanges = TestDataGenerator.PriceBars(TestDataGenerator.NET) |> PercentChangeAnalysis.calculateForPriceBars true

    [<Fact>]
    let ``Mean is correct``() = Assert.Equal(0.05m, percentChanges.mean, 2)

    [<Fact>]
    let ``StdDev is correct``() = Assert.Equal(5.21m, percentChanges.stdDev, 2)

    [<Fact>]
    let ``Min is correct``() = Assert.Equal(-18.42m, percentChanges.min)

    [<Fact>]
    let ``Max is correct``() = Assert.Equal(27.06m, percentChanges.max)

    [<Fact>]
    let ``Median is correct``() = Assert.Equal(0.17m, percentChanges.median)

    [<Fact>]
    let ``Skewness is correct``() = Assert.Equal(0.36m, percentChanges.skewness, 2)

    [<Fact>]
    let ``Kurtosis is correct``() = Assert.Equal(2.89m, percentChanges.kurtosis, 2)

    [<Fact>]
    let ``Buckets are correct``() =
        Assert.Equal(21, percentChanges.buckets.Length)
        
        Assert.Equal(18, percentChanges.buckets |> Array.filter (fun x -> x.frequency <> 0) |> Array.length)
        
        // last bucket should include max
        let lastBucketValue = percentChanges.buckets |> Array.last |> _.value
        
        Assert.True(percentChanges.max > lastBucketValue)
        Assert.Equal(1, percentChanges.buckets |> Array.last |> _.frequency)
